_contract-shared-headers/eosdactokens_shared.hpp

```diff
@@ -35,23 +40,113 @@ namespace eosdac {
         uint64_t primary_key() const { return balance.symbol.code().raw(); }
     };
 
+    TABLE currency_stats {
+            asset supply;
+            asset max_supply;
+            name issuer;
+            bool transfer_locked = false;
+
+            uint64_t primary_key() const { return supply.symbol.code().raw(); }
+    };
+
+    typedef eosio::multi_index<"stat"_n, currency_stats> stats;
     typedef eosio::multi_index<"members"_n, member> regmembers;
     typedef eosio::multi_index<"accounts"_n, account> accounts;
 
 
-    static void assertValidMember(eosio::name member, eosio::name dac_scope) {
+    TABLE stake_info {
+            name  account;
+            asset stake;
+
+            uint64_t primary_key() const { return account.value; }
+    };
+    typedef multi_index<"stakes"_n, stake_info> stakes_table;
+
+    TABLE unstake_info {
+            uint64_t       key;
+            name           account;
+            asset          stake;
+            time_point_sec release_time;
+
+            uint64_t primary_key() const { return key; }
+            uint64_t by_account() const { return account.value; }
+    };
+    typedef multi_index<"unstakes"_n, unstake_info,
+            indexed_by<"byaccount"_n, const_mem_fun<unstake_info, uint64_t, &unstake_info::by_account> >
+    > unstakes_table;
+
+    TABLE staketime_info {
+            name           account;
+            uint32_t       delay;
+
+            uint64_t primary_key() const { return account.value; }
+    };
+    typedef multi_index<"staketime"_n, staketime_info > staketimes_table;
+
+    asset get_supply(name code, symbol_code sym) {
+        stats statstable(code, sym.raw());
+        const auto &st = statstable.get(sym.raw());
+        return st.supply;
+    }
+
+    asset get_balance(name owner, name code, symbol_code sym) {
+        accounts accountstable(code, owner.value);
+        const auto &ac = accountstable.get(sym.raw());
+        return ac.balance;
+    }
+
+    asset get_liquid(name owner, name code, symbol_code sym) {
```

Comment:
I prefer the option of keeping the balance reflecting the actual balance.  There are pros and cons to each method, the con of transferring the tokens is that the user cannot see that the tokens exist at all without a special workaround.  By blocking transfers, we can at least provide a meaningful message to users and our member client will always display the staked and liquid balances.

Yes it is true that wallets may attempt to transfer staked tokens, but since the user has to manually stake, it is likely that they will be aware of the fact that they are staked.  I think actually displaying their full balance is more important.  When DAC's rule the world, wallets will be forced to follow our staking standard anyway

Reply:
So true. Wallets will follow. The innovative ones will follow soon. 😄